File: src/js/src/tp.ts
import { HEADER_SIZE } from './codec.js';

export class TpHeader {
    constructor(
        public offset: number, // In units of 16 bytes
        public moreSegments: boolean
    ) { }

    static deserialize(buf: Buffer): TpHeader {
        if (buf.length < 4) throw new Error("TP Header too short");
        const val = buf.readUInt32BE(0);
        const offset = val >>> 4;
        const moreSegments = (val & 0x01) !== 0;
        return new TpHeader(offset, moreSegments);
    }

    serialize(): Buffer {
        const buf = Buffer.alloc(4);
        // Offset is 28 bits, shifted left by 4. More segments is last bit.
        const val = (this.offset << 4) | (this.moreSegments ? 1 : 0);
        buf.writeUInt32BE(val >>> 0, 0); // Ensure unsigned
        return buf;
    }
}

/**
 * Splits a payload into TP segments.
 * @param payload The full payload to split.
 * @param maxPayloadPerSegment Max bytes per segment (must be multiple of 16).
 * @returns Array of [TpHeader, Buffer] tuples.
 */
export function segmentPayload(payload: Buffer, maxPayloadPerSegment: number = 1392): { header: TpHeader, chunk: Buffer }[] {
    const segments: { header: TpHeader, chunk: Buffer }[] = [];
    let remaining = payload.length;
    let currentOffset = 0; // In bytes

    while (remaining > 0) {
        let chunkSize = Math.min(remaining, maxPayloadPerSegment);

        // If not the last segment, length must be multiple of 16
        if (remaining > maxPayloadPerSegment) {
            chunkSize = Math.floor(chunkSize / 16) * 16;
        }

        const chunk = payload.subarray(currentOffset, currentOffset + chunkSize);
        remaining -= chunkSize;

        const moreSegments = remaining > 0;
        // Offset in TpHeader is in units of 16 bytes
        const tpOffset = currentOffset / 16;

        segments.push({
            header: new TpHeader(tpOffset, moreSegments),
            chunk: Buffer.from(chunk) // Copy to be safe
        });

        currentOffset += chunkSize;
    }

    return segments;
}

export class TpReassembler {
    // Key: "serviceId:methodId:clientId:sessionId"
    private buffers = new Map<string, {
        segments: Map<number, Buffer>, // offset -> data
        lastOffsetReceived: boolean,
        expectedTotalLength: number
    }>();

    private MAX_PAYLOAD_SIZE = 10 * 1024 * 1024; // 10MB limit

    /**
     * Process a TP segment.
     * @returns The full reassembled payload if complete, null otherwise.
     */
    processSegment(
        key: { serviceId: number, methodId: number, clientId: number, sessionId: number },
        header: TpHeader,
        payload: Buffer
    ): Buffer | null {
        const k = `${key.serviceId}:${key.methodId}:${key.clientId}:${key.sessionId}`;

        let session = this.buffers.get(k);
        if (!session) {
            session = { segments: new Map(), lastOffsetReceived: false, expectedTotalLength: 0 };
            this.buffers.set(k, session);
        }

        // Validate payload length alignment (must be multiple of 16 unless it's the last one)
        // Standard says: "The length of the payload of a segment... shall be a multiple of 16 bytes.
        // Exception: The last segment..."
        if (header.moreSegments && (payload.length % 16 !== 0)) {
            console.warn(`[TP] Dropping invalid segment: More=1 but len=${payload.length} not aligned.`);
            this.buffers.delete(k);
            return null;
        }

        session.segments.set(header.offset, payload);

        if (!header.moreSegments) {
            session.lastOffsetReceived = true;
            // The byte offset of this last chunk starts at header.offset * 16
            // Total length = (header.offset * 16) + payload.length
            session.expectedTotalLength = (header.offset * 16) + payload.length;
        }

        // Check completion
        if (session.lastOffsetReceived) {
            // Check if we have all bytes covered
            let currentOffset = 0;
            let totalLen = 0;
            // Simplified check: Use a sorted approach or just track holes?
            // Since we stored by unit-offset, we can iterate.
            // But Map iteration order is insertion order in JS (mostly).
            // Let's sort keys.
            const sortedOffsets = Array.from(session.segments.keys()).sort((a, b) => a - b);

            // Check contiguity
            // First offset must be 0?
            if (sortedOffsets.length === 0 || sortedOffsets[0] !== 0) return null;

            let contiguous = true;
            let byteOffset = 0;

            for (const off of sortedOffsets) {
                if (off * 16 !== byteOffset) {
                    contiguous = false;
                    break;
                }
                const chunk = session.segments.get(off)!;
                byteOffset += chunk.length;
            }

            if (contiguous && byteOffset === session.expectedTotalLength) {
                // Reassemble
                const chunks = sortedOffsets.map(off => session.segments.get(off)!);
                const full = Buffer.concat(chunks);
                this.buffers.delete(k);
                return full;
            }
        }

        return null;
    }
}

File: src/cpp/src/tp.cpp
#include "fusion_hawking/tp.hpp"
#include <sstream>
#include <iostream>
#include <algorithm>
#include <cstring>

namespace fusion_hawking {

std::vector<uint8_t> TpHeader::serialize() const {
    std::vector<uint8_t> buf(4);
    uint32_t val = (offset << 4) | (more_segments ? 1 : 0);
    buf[0] = (val >> 24) & 0xFF;
    buf[1] = (val >> 16) & 0xFF;
    buf[2] = (val >> 8) & 0xFF;
    buf[3] = val & 0xFF;
    return buf;
}

TpHeader TpHeader::deserialize(const uint8_t* buffer, size_t len) {
    if (len < 4) return {0, false};
    uint32_t val = (buffer[0] << 24) | (buffer[1] << 16) | (buffer[2] << 8) | buffer[3];
    TpHeader h;
    h.offset = val >> 4;
    h.more_segments = (val & 0x01) != 0;
    return h;
}

bool TpReassembler::process_segment(uint16_t service_id, uint16_t method_id, uint16_t client_id, uint16_t session_id, const TpHeader& header, const std::vector<uint8_t>& payload, std::vector<uint8_t>& out_payload) {
    std::stringstream key_ss;
    key_ss << service_id << ":" << method_id << ":" << client_id << ":" << session_id;
    std::string key = key_ss.str();

    auto& session = buffers[key];

    // Check alignment
    if (header.more_segments && (payload.size() % 16 != 0)) {
        std::cerr << "[TP] Dropping invalid segment: More=1 but len=" << payload.size() << " not aligned." << std::endl;
        buffers.erase(key);
        return false;
    }

    session.segments[header.offset] = payload;

    if (!header.more_segments) {
        session.last_offset_received = true;
        session.expected_total_length = (header.offset * 16) + payload.size();
    }

    if (session.last_offset_received) {
        // Check continuity
        uint32_t current_offset = 0; // expected offset 0
        uint32_t byte_offset = 0;
        
        bool contiguous = true;
        
        // Map is sorted by key (offset)
        if (session.segments.empty() || session.segments.begin()->first != 0) return false;

        for (auto const& [off, chunk] : session.segments) {
             // 'off' is in 16-byte units.
             // If byte_offset (bytes received so far) != off * 16, then gap.
             if (off * 16 != byte_offset) {
                 contiguous = false;
                 break;
             }
             byte_offset += chunk.size();
        }

        if (contiguous && byte_offset == session.expected_total_length) {
            // Reassemble
            out_payload.clear();
            out_payload.reserve(session.expected_total_length);
            for (auto const& [off, chunk] : session.segments) {
                out_payload.insert(out_payload.end(), chunk.begin(), chunk.end());
            }
            buffers.erase(key);
            return true;
        }
    }
    return false;
}

std::vector<std::pair<TpHeader, std::vector<uint8_t>>> segment_payload(const std::vector<uint8_t>& payload, uint32_t max_segment_size) {
    std::vector<std::pair<TpHeader, std::vector<uint8_t>>> segments;
    size_t total_len = payload.size();
    size_t current_pos = 0; // bytes

    while (current_pos < total_len) {
        size_t remaining = total_len - current_pos;
        size_t chunk_size = std::min((size_t)max_segment_size, remaining);

        // Align if not last
        bool more = false;
        if (remaining > max_segment_size) {
            chunk_size = (chunk_size / 16) * 16;
            more = true;
        } else {
             more = false;
        }

        TpHeader h;
        h.offset = (uint32_t)(current_pos / 16);
        h.more_segments = more;

        std::vector<uint8_t> chunk(payload.begin() + current_pos, payload.begin() + current_pos + chunk_size);
        segments.push_back({h, chunk});
        
        current_pos += chunk_size;
    }
    return segments;
}

}
